File: CS50X/Week 2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int countLetters(string text);
int countWords(string text);
int countSentences(string text);

int main(void)
{
    // Prompt the user for some text
    string text = get_string("Text: ");

    // Count the number of letters, words, and sentences in the text
    int letters = countLetters(text);
    int words = countWords(text);
    int sentences = countSentences(text);

    // Compute the Coleman-Liau index
    float L = (float) letters * 100 / words;
    float S = (float) sentences * 100 / words;
    float index = 0.0588 * L - 0.296 * S - 15.8;

    // Print the grade level
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        int grade = round(index);
        if (grade == 1)
        {
            printf("Grade 1\n");
        }
        else if (grade >= 16)
        {
            printf("Grade 16+\n");
        }
        else
        {
            printf("Grade %i\n", grade);
        }
    }
}

int countLetters(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            count++;
        }
    }

    return count;
}

int countWords(string text)
{
    int count = 1;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isspace(text[i]))
        {
            count++;
        }
    }

    return count;
}

int countSentences(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count++;
        }
    }
    
    return count;
}

File: CS50X/Week 6/sentimental-readability/readability.py
import re


def count_letters(text):
    # Count letters by filtering out non-letter characters and counting the rest
    letters = re.findall(r'[a-zA-Z]', text)
    return len(letters)


def count_words(text):
    # Count words by splitting the text by whitespace
    words = text.split()
    return len(words)


def count_sentences(text):
    # Count sentences by counting ., !, and ?
    sentences = re.findall(r'[.!?]', text)
    return len(sentences)


def coleman_liau_index(text):
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # Calculate L and S
    L = letters / words * 100
    S = sentences / words * 100

    # Calculate the Coleman-Liau index
    index = 0.0588 * L - 0.296 * S - 15.8
    return index


def main():
    # Get user input
    text = input("Text: ")

    # Calculate the readability grade
    index = round(coleman_liau_index(text))

    # Print the grade level
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


if __name__ == "__main__":
    main()
